File: hihocoder.1105.heap/Main.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Queue<Integer> heap = new PriorityQueue<>(Comparator.reverseOrder());

        Scanner in = new Scanner(System.in);
        int N = in.nextInt();
        for (int i = 0; i < N; i++) {
            String operation = in.next();
            if (operation.equals("A"))
                heap.offer(in.nextInt());
            else
                System.out.println(heap.poll());
        }
    }
}

File: hihocoder.1105.heap/main.cpp
#include <iostream>
#include <queue>
using namespace std;

int main()
{
	priority_queue<int> heap;

	int N;
	cin >> N;
	for (int i = 0; i < N; i++)
	{
		char operation;
		cin >> operation;
		if (operation == 'A')
		{
			int W;
			cin >> W;
			heap.push(W);
		}
		else
		{
			cout << heap.top() << endl;
			heap.pop();
		}
	}
}
